net/bonding/files/lib/netifd/proto/bonding.sh

```diff
@@ -0,0 +1,237 @@
+		;;
+        esac
+
+	# Set link monitoring (with corresponding parameters)
+	local link_monitoring
+	json_get_vars link_monitoring
+
+	case "$link_monitoring" in
+
+		arp)
+			local arp_interval arp_ip_target arp_all_targets arp_validate
+			json_get_vars arp_interval arp_ip_target arp_all_targets arp_validate
+
+			[ -n "$arp_interval" -a "$arp_interval" != 0 ] && echo "$arp_interval" > /sys/class/net/"$link"/bonding/arp_interval
+
+			IFS=' '
+			for target in $arp_ip_target; do
+				echo "+$target" > /sys/class/net/"$link"/bonding/arp_ip_target
+			done
+
+			[ -n "$arp_all_targets" ] && echo "$arp_all_targets" > /sys/class/net/"$link"/bonding/arp_all_targets
+			[ -n "$arp_validate" ] && echo "$arp_validate" > /sys/class/net/"$link"/bonding/arp_validate
+		;;
+
+		mii)
+			local miimon downdelay updelay use_carrier
+			json_get_vars miimon downdelay updelay use_carrier
+
+			[ -n "$miimon" -a "$miimon" != 0 ] && echo "$miimon" > /sys/class/net/"$link"/bonding/miimon
+			[ -n "$downdelay" ] && echo "$downdelay" > /sys/class/net/"$link"/bonding/downdelay
+			[ -n "$updelay" ] && echo "$updelay" > /sys/class/net/"$link"/bonding/updelay
+			[ -n "$use_carrier" ] && echo "$use_carrier" > /sys/class/net/"$link"/bonding/use_carrier
+		;;
+	esac
+
+	# Add slaves to bonding interface
+	local slaves
+	json_get_vars slaves
+
+	for slave in $slaves; do
+
+		if [ "$(cat /proc/net/dev |grep "$slave")" == "" ]; then
+			echo "$cfg" "ERROR IN CONFIGURATION - $slave: No such device"
+			proto_notify_error "$cfg" "ERROR IN CONFIGURATION - $slave: No such device"
+			proto_block_restart "$cfg"
+			return
+		fi
+
+		ifconfig "$slave" down
+
+		sleep 1
+
+		echo "+$slave" > /sys/class/net/"$link"/bonding/slaves
+
+		ifconfig "$slave" up
+	done
+
+	[ -n "$all_slaves_active" ] && echo "$all_slaves_active" > /sys/class/net/"$link"/bonding/all_slaves_active
+
+	# Check for static or DHCP configuration
+	local dhcp
+	json_get_vars dhcp
+
+	local ipaddr netmask
+	json_get_vars ipaddr netmask
+
+	# ATTENTION
+	#All json vars have to be read before the line below, as the
+	# json object will be overwritten by proto_init_update
+	# ATTENTION
+
+	proto_init_update "$link" 1
+
+	# For static configuration we _MUST_ have an IP address
+	[ -z "$ipaddr" ] && {
+		echo "$cfg" "MISSING STATIC IP ADDRESS"
+		proto_notify_error "$cfg" "MISSING STATIC IP ADDRESS"
+		proto_block_restart "$cfg"
+		return
+	}
+
+	proto_add_ipv4_address "$ipaddr" "$netmask"
+
+	proto_send_update "$cfg"
+}
+
+proto_bonding_teardown() {
+	local cfg="$1"
+	local link="bonding-$cfg"
+
+	# Check for loaded kernel bonding driver (/sys/class/net/bonding_masters exists)
+	[ -f "$BONDING_MASTERS" ] || {
+		echo "$cfg" "teardown: bonding_masters does not exist in sysfs (kernel module not loaded?)"
+		proto_notify_error "$cfg" "teardown: bonding_masters does not exist in sysfs (kernel module not loaded?)"
+		proto_block_restart "$cfg"
+		return
+	}
```

Comment:
Not sure how much sense it makes to check the presence of the module on teardown, at least I'd probably make it silent to not double the amount of reported errors in case of an existing config but missing kmod.

But I don't have an overly strong opinion on it.

Reply:
Well, there's only one line in log if the kernel module gets unloaded during operation.
I think it's a good advice in the log to catch the problem.
Let's have it unchanged, OK?